# Naive Bayes

## Estudo 03 - Classificação Base Censu

In [31]:
#Na tabela estudada 1: paga emprestimo e 0: nao paga
previsoes

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [33]:
#devemos comparar com y_credit_test, vamos ver se o modelo esta funcionando
y_credit_teste

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [54]:
# Importar agora a bliblioteca que analisa as métricas

from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

In [40]:
# Chamar a funcao que recebe 2 parametros, y_true -> repostas reais e y_predict -> o que o algoritimo ta falando 

accuracy_score(y_credit_teste,previsoes)

0.938

In [1]:
# Importando as Blibliotecas

import pandas as pd

from sklearn.preprocessing import LabelEncoder

import pickle as pk

In [6]:
#Abrindo o arquivo que vamos estudar que ja passou por um Pré-Processamento
with open(r'content\03.census\census.pkl', 'rb') as f:
    X_census_treinamento, y_census_treinamento, X_census_teste, y_census_teste = pk.load(f)

In [8]:
#Tamanho de cada x so 108 variaveis e y se a pessoa ganha 50k dolares ou nao por mes
X_census_treinamento.shape, y_census_treinamento.shape

((27676, 108), (27676,))

In [11]:
X_census_teste.shape, y_census_teste.shape

((4885, 108), (4885,))

In [13]:
#Importa a Bliblioteca para treinamento do Naive Bayes

from sklearn.naive_bayes import GaussianNB

In [16]:
#Instancia o modelo 
naive_census = GaussianNB()
# Realiza o treinamento
naive_census.fit(X_census_treinamento, y_census_treinamento)
#Gera as previsoes com os dados de teste
previsoes = naive_census.predict(X_census_teste)
previsoes

array([' >50K', ' >50K', ' >50K', ..., ' <=50K', ' >50K', ' >50K'],
      shape=(4885,), dtype='<U6')

In [17]:
y_census_teste

array([' <=50K', ' <=50K', ' <=50K', ..., ' <=50K', ' <=50K', ' <=50K'],
      shape=(4885,), dtype=object)

In [22]:
#Importa a Bliblioteca de Métricas 

from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

In [24]:
accuracy_score(y_census_teste,previsoes)

0.4767656090071648

In [25]:
# Era melhor jogar uma moeda cara ou coroa que voce acertaria mais.
# Preciso fazer testes para melhorar

In [26]:
confusion_matrix(y_census_teste,previsoes)

array([[1172, 2521],
       [  35, 1157]])

In [30]:
naive_census.classes_

array([' <=50K', ' >50K'], dtype='<U6')

In [29]:
print(classification_report(y_census_teste,previsoes))

              precision    recall  f1-score   support

       <=50K       0.97      0.32      0.48      3693
        >50K       0.31      0.97      0.48      1192

    accuracy                           0.48      4885
   macro avg       0.64      0.64      0.48      4885
weighted avg       0.81      0.48      0.48      4885



## 📊 Análise do Modelo Gaussian Naive Bayes - Census Income

### 📌 Distribuição dos dados de teste (y_teste)

* 75,6% dos indivíduos possuem renda menor ou igual a 50 mil dólares (3.693 de 4.885)
* 24,4% dos indivíduos possuem renda superior a 50 mil dólares (1.192 de 4.885)

Isso demonstra um cenário desbalanceado, onde a maior parte da população pertence à faixa de renda até 50 mil dólares.

---

## 🧠 Matriz de Confusão

| Real \ Previsto | <=50K | >50K |
| --------------- | ----- | ---- |
| <=50K           | 1172  | 2521 |
| >50K            | 35    | 1157 |

---

## ✅ Classe <=50K — Renda até 50 mil dólares

O modelo apresentou um comportamento bastante conservador ao realizar previsões dessa classe.

* **Precision: 97%**

Quando o modelo prevê que uma pessoa possui renda de até 50 mil dólares, ele está correto na grande maioria das vezes.

* **Recall: 32%**

O modelo conseguiu identificar apenas 32% das pessoas que realmente pertencem a essa faixa de renda.

📌 **Interpretação:**

Embora as previsões sejam altamente confiáveis, o modelo deixa de reconhecer grande parte dos indivíduos dessa classe.

➡️ Aproximadamente 68% das pessoas que realmente possuem renda até 50 mil dólares foram classificadas incorretamente como pertencentes à faixa de renda superior.

---

## ⚠️ Classe >50K — Renda superior a 50 mil dólares

Esta foi a classe favorecida pelo modelo durante as previsões.

* **Precision: 31%**

Quando o modelo prevê que uma pessoa possui renda superior a 50 mil dólares, ele acerta apenas 31% das vezes.

* **Recall: 97%**

Por outro lado, o modelo consegue encontrar praticamente todas as pessoas que realmente pertencem a essa faixa de renda.

📌 **Interpretação:**

O modelo identifica quase todos os indivíduos de alta renda, porém gera uma grande quantidade de falsos positivos.

➡️ Muitas pessoas com renda inferior a 50 mil dólares são classificadas incorretamente como pertencentes à faixa de renda superior.

---

## 📈 Visão Geral do Modelo

* **F1-Score (<=50K): 48%**
* **F1-Score (>50K): 48%**
* **Accuracy Geral: 48%**

O F1-Score evidencia que o modelo não conseguiu equilibrar adequadamente precisão e recall para nenhuma das classes.

Embora o recall da classe >50K seja muito elevado, a baixa precisão reduz significativamente a qualidade das previsões.

Da mesma forma, a classe <=50K apresenta excelente precisão, mas um recall extremamente baixo.

---

## 🧾 Conclusão

O Gaussian Naive Bayes apresentou forte desequilíbrio entre as classes analisadas.

O modelo demonstrou excelente capacidade de encontrar indivíduos com renda superior a 50 mil dólares, porém ao custo de classificar incorretamente um grande número de pessoas da faixa de renda inferior.

📌 Em termos práticos:

* O modelo dificilmente deixa de identificar indivíduos de alta renda 👍
* Porém classifica muitas pessoas de renda mais baixa como pertencentes à faixa de alta renda ⚠️

Como este estudo teve foco educacional, os resultados foram importantes para compreender o funcionamento do algoritmo Naive Bayes e a interpretação de métricas como Accuracy, Precision, Recall, F1-Score e Matriz de Confusão.

➡️ Trabalhos futuros podem explorar engenharia de atributos, balanceamento de classes e comparação com outros algoritmos de classificação para melhorar a capacidade de generalização do modelo.
